docs/proposals/2841-flux-hpc/README.md

```diff
@@ -0,0 +1,185 @@
+5.  **Networking**: The plugin ensures the `JobSet` is configured with a headless service and DNS hostnames enabled, which Flux uses for its broker communication. High speed network for MPI can be used by way of extending pods to use a bypass mechanism to support Infiniband (Azure) or the Elastic Fabric Adapter (AWS).
+
+This approach provides an HPC environment without requiring the user to build a custom Docker image with Flux pre-installed, significantly improving the user experience.
+
+### API Design
+
+The proposed changes will integrate into the existing `v1alpha1` API structures. We will add a new field, `flux`, to the `MLPolicySource` struct. This aligns Flux with other runtimes like Torch and MPI.
+
+#### 1. `ClusterTrainingRuntime` and `TrainingRuntime`
+
+The proposal will leverage the existing `ClusterTrainingRuntime` (cluster-scoped) and `TrainingRuntime` (namespace-scoped) CRDs. These objects serve as reusable templates, and no changes are needed for their top-level structure.
+
+#### 2. Enhancing `MLPolicySource`
+
+We will add the `Flux` field to the `MLPolicySource` struct, which is embedded within the `MLPolicy`.
+
+```go
+type MLPolicySource struct {
+    Torch *TorchMLPolicySource `json:"torch,omitempty"`
+    MPI   *MPIMLPolicySource   `json:"mpi,omitempty"`
+
+    // FluxMLPolicy defines policy only for Flux
+    // +optional
+    Flux  *FluxMLPolicySource  `json:"flux,omitempty"`
+}
+```
+
+For the `FluxMLPolicySource`, we define the minimum required parameters needed for Flux and installing the view. The view must be compatible with the application container.
+
+```go
+// FluxMLPolicySource represents a Flux HPC runtime configuration.
+type FluxMLPolicySource struct {
+
+    // numNodes is the number of physical nodes for the job.
+    // This is defined a level up on the Trainer
+
+    // numProcPerNode is the number of processes per node.
+    // Defaults to 1.
+    // +kubebuilder:default=1
+    // +optional
+    NumProcPerNode *int32 `json:"numProcPerNode,omitempty"`
+
+    // Container image to use for Flux view that installs Flux
+    // This must be compatible with the application container
+    // Get the flux view container (these are choices)
+    // ghcr.io/converged-computing/flux-view-rocky:arm-9
+    // ghcr.io/converged-computing/flux-view-rocky:arn-8
+    // ghcr.io/converged-computing/flux-view-rocky:tag-9
+    // ghcr.io/converged-computing/flux-view-rocky:tag-8
+    // ghcr.io/converged-computing/flux-view-ubuntu:tag-noble
+    // ghcr.io/converged-computing/flux-view-ubuntu:tag-jammy
+    // ghcr.io/converged-computing/flux-view-ubuntu:tag-focal
+    // ghcr.io/converged-computing/flux-view-ubuntu:arm-jammy
+    // ghcr.io/converged-computing/flux-view-ubuntu:arm-focal
+    // We use an ubuntu (more recent) default since it is common
+    // +kubebuilder:default="ghcr.io/converged-computing/flux-view-ubuntu:arm-jammy"
+    Image string `json:"image,omitempty"`
```

Comment:
Can we directly add the image in the ClusterTrainingRuntime's initContainer spec? That will allow users to customize other specification for the init container if they want (e.g. env variables, volumes, etc.), and also remove number of parameters from FluxMLPolicySource API.

WDYT @vsoch ?

Reply:
Oh, yes I think we could try that. If it doesn't work I can report back and we add this abck. I'll remove it for now.

Reply:
@vsoch Please can you also update the `flux-runtime` spec to include this InitContainer example?